### ISSUE TITLE
Cpp opencv4.7.0 could not load exported ONNX model

### ISSUE THREAD

**Issue by qqqhhh-any** [ultralytics/ultralytics#2198]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### YOLOv8 Component

Integrations

### Bug

Hello,I exported yolov8n model to onnx ``` yolo mode=export model=best.pt opset=12 format=onnx ```, and load it by 
```net = cv::dnn::readNetFromONNX(modelPath);```
unlucky,it dose not work.The error code is:
```
[ERROR:0@1.711] global onnx_importer.cpp:1054 cv::dnn::dnn4_v20221220::ONNXImporter::handleNode DNN/ONNX: ERROR during processing node with 2 inputs and 1 outputs: [Reshape]:(onnx_node!Reshape_342) from domain='ai.onnx'
```
I followed the instructions from <https://github.com/opencv/opencv/issues/23174> to simplify the model,but it still not work.

### Environment

Ultralytics YOLOv8.0.85 🚀 Python-3.9 torch-1.12.0+cu116 CUDA:11.6 (NVIDIA GeForce GTX 3060 Laptop)
Windows 10 
opencv 4.7.0
Python3.9

### Minimal Reproducible Example

export model
```
yolo mode=export model=best.pt opset=12 format=onnx 
```
load model in cpp
```
net = cv::dnn::readNetFromONNX(modelPath);
```

### Are you willing to submit a PR?

- [X] Yes I'd like to help by submitting a PR!

**Comment by qqqhhh-any** [comment]
```
from ultralytics import YOLO

# Load a model
model = YOLO("yolov8x.yaml")  # build a new model from scratch
# model = YOLO("yolov8n.pt")  # load a pretrained model (recommended for training)

# Use the model
model.train(data="coco128.yaml", epochs=2)  # train the model
metrics = model.val()  # evaluate model performance on the validation set
results = model("https://ultralytics.com/images/bus.jpg")  # predict on an image
success = model.export(format="onnx")  # export the model to ONNX format
```
follow the code to export model,it can be loaded successfully.
It seems that use onnxsim to optimize the onnx model can fix the bug

**Comment by glenn-jocher** [comment]
@qqqhhh-any thank you for sharing your experience. It appears that you were able to successfully export your YOLOv8 model to ONNX format and load it in your application. However, you encountered an error when trying to use the loaded model. The error code is related to a `Reshape` operation and may occur when the ONNX model is not compliant with Opset 12.

One possible solution is to use `onnxsim` to optimize the ONNX model, which may fix the problem. Additionally, we recommend referring to the official OpenCV documentation for more guidance on using ONNX models with `cv::dnn::readNetFromONNX()`.

**Comment by shahla-ai** [comment]
Hello @glenn-jocher 
I faced the same problem, i converted yolov8n.pt into the onnx format using the command `yolo export model=yolov8n.pt format=onnx opset=12`
I'm trying to run the **yolov8-cpp-inference** in the **examples** folder in **ultralytics**, but i had the same issue mentioned above
Any help ?
Best Regards

**Comment by glenn-jocher** [comment]
@shahla-ai hi there,

If you've converted your `yolov8n.pt` to onnx format following the recommended steps and you're facing issues while running the yolov8-cpp-inference, here's a couple of things you might want to check:

1. Ensure you're using a supported and  compatible version of OpenCV with onnx for the C++ environment. You may want to update or roll-back your OpenCV version to check if this resolves the issue.

2. Check the opset version you're using in the export command. Different opset versions may contain operations not supported by C++ onnx runtime.

Lastly, when you are loading your ONNX model in C++, ensure that the path to the ONNX file is correct. If your directory structure doesn't match what's expected by the yolov8-cpp-inference example, it could lead to issues as well.

I hope these suggestions help you in resolving the issue. And remember, you can always reach out to the Ultralytics team and the YOLO community if you need further assistance.

Best Regards!

**Comment by shahla-ai** [comment]
Thank you @glenn-jocher 
I rolled back my OpenCV to version 4.5.4 built with cuda and onnx support successfully, but i faced the same problem
I used different opset vesrions(12,14,17) in my export process but with no change
the exact error is `[ERROR:0] global /home/pc/Downloads/OpenCV/opencv-4.5.4/modules/dnn/src/onnx/onnx_importer.cpp (720) handleNode DNN/ONNX: ERROR during processing node with 2 inputs and 1 outputs: [Reshape]:(/model.22/dfl/Reshape_output_0)`
Any extra help please

**Comment by glenn-jocher** [comment]
@shahla-ai i'm sorry that you're still having issues using ONNX models with OpenCV.

Since you've mentioned that updating OpenCV to version 4.5.4 and trying multiple opset versions hasn't solved the problem, the error might be arising due to Reshape layer issue in YOLOv8 transformed ONNX model.

Without modifying your underlying model or the OpenCV library, you might want to try reducing ONNX opset version further or try exporting the model with a lower ONNX version or a different format. 

It's also worth noting that OpenCV's DNN module doesn’t fully support all the ONNX operations or sometimes fails to load the ONNX models correctly.

I hope this suggestion will be useful to you, and I look forward to hearing about your resolution soon.

**Comment by shahla-ai** [comment]
@glenn-jocher 
First of all i would like to thank you for your replies
The thing is that i tried to run **yolov8-cpp-inference** to reach the maximum performance of a **yolov8 .pt model**
I will be very thankful if you guided me to an another approach to reach the highest performance speed without dropping the accuracy 
I'm working with the **yolov8 nano model**, i tried the default **CLI** inference, the **Python ONNXRuntime** inference, and the **Python ONNX OpenCV** inference and all worked well but i need to know the approach that gives me the highest speed at inference **even if it's not one of those approaches**
Thanks again for your help
Best Regards

**Comment by glenn-jocher** [comment]
@shahla-ai Thank you for your engagement and persistence in optimizing the performance of the YOLOv8 nano model. 

To reach maximum inference performance, it's important to leverage hardware acceleration whenever possible. That being said, if you're running on an NVIDIA GPU, you might want to consider using TensorRT. TensorRT is a deep learning model optimizer and runtime library developed by NVIDIA for GPU acceleration. It's known for being able to significantly speed up inference times.

Aside from TensorRT, many find good performance improvements using ONNX Runtime with a properly configured execution provider, depending on your specific hardware. This could be the CUDA execution provider for NVIDIA GPUs, or other execution providers for different hardware.

One more approach would be to try and optimize the model itself, for example through quantization or pruning, though these techniques usually come with some loss of accuracy.

Remember, the right approach heavily depends on your specific hardware and software environment, and it may require some experimentation to find the method that provides the optimal balance between speed and accuracy for your specific use case.

Best regards!
